design/include-cleaner.md

```diff
@@ -0,0 +1,187 @@
+# Include Cleaner
+
+Manually managing includes in a C++ translation unit, especially in the face of
+transitive inclusions, requires a lot of effort. Include Cleaner aims to
+provide diagnostics to keep includes in an
+[IWYU](https://include-what-you-use.org/)-clean state.
+
+Include Cleaner is available in "preview" mode with an incomplete set of
+capabilities and can be enabled through [configuration
+file](/config#UnusedIncludes). If you experience any bugs, please submit a bug
+report in [clangd/issues](https://github.com/clangd/clangd/issues).
+
+{:.v14}
+
+## How it works
+
+Include Cleaner issues diagnostics for includes that are present but not used
+in the main file. When you open a file, clangd will analyze the symbols the
+file uses and mark all headers defining these symbols as "used". The warnings
+will be issued for the headers that are included but not marked as "used".
+Example:
+
+```c++
+// foo.h
+class Foo {};
+```
+
+```c++
+// bar.h
+class Bar {};
+```
+
+```c++
+// main.cpp
+
+#include "foo.h"
+#include "bar.h" // <- Will be marked as unused and suggested to be removed.
+
+int main() {
+  Foo f;
+}
+```
+
+Here, `main.cpp` only makes use of symbols from `foo.h` and removing `#include
+"bar.h"` prevents unnecessary parsing of `bar.h` and allows breaking the
+dependency on it.
+
+### Deciding what headers are used
+
+Clangd relies on Clang AST to deduce which headers are used and which aren't,
+the whole Include Cleaner decision process is described below.
+
+#### Scanning the main file
+
+IncludeCleaner will traverse Clang AST of the main file (the file currently
```

Comment:
i am not sure `the file currently opened in the editor` really applies at this layer. I'd probably drop that, if you really want to give description of a main file we should probably go with `Part of a source file after first chunk of includes and other pre-processor directives (called preamble)`.